From bb8eacb2fa511c56d87b83581d2e35ed1e45dcdb Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Tue, 16 Jan 2024 12:14:28 +0530
Subject: [PATCH] wifi: mac80211: fix race condition on enabling fast-xmit

fast-xmit must only be enabled after the sta has been uploaded to
the driver, otherwise it could end up passing the not-yet-uploaded
sta via drv_tx calls to the driver, leading to potential crashes
because of uninitialized drv_priv data.

Add a missing sta->uploaded check and re-check fast xmit after
inserting a sta.

Signed-off-by: Felix Fietkau <nbd@nbd.name>
Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 net/mac80211/sta_info.c | 1 +
 net/mac80211/tx.c       | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/sta_info.c b/net/mac80211/sta_info.c
index 7a32949..7196edf 100644
--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -945,6 +945,7 @@ static int sta_info_insert_finish(struct sta_info *sta) __acquires(RCU)
 
 	if (ieee80211_vif_is_mesh(&sdata->vif))
 		mesh_accept_plinks_update(sdata);
+	ieee80211_check_fast_xmit(sta);
 
 	return 0;
  out_remove:
diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index aca9e8c..1983f7d 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3103,7 +3103,7 @@ void ieee80211_check_fast_xmit(struct sta_info *sta)
 	    sdata->vif.type == NL80211_IFTYPE_STATION)
 		goto out;
 
-	if (!test_sta_flag(sta, WLAN_STA_AUTHORIZED))
+	if (!test_sta_flag(sta, WLAN_STA_AUTHORIZED) || !sta->uploaded)
 		goto out;
 
 	if (test_sta_flag(sta, WLAN_STA_PS_STA) ||
-- 
2.34.1

